### ISSUE TITLE
UTF8 character uses wrong width for text wrapping

### ISSUE THREAD

**Issue by murk-sy** [dompdf/dompdf#1124]
Using 0.7.0-beta 3, DejaVu Sans font - characters themselves work as they should. Tested on PHP 5.3.1 and PHP 5.5.30. Dompdf is unmodified.

When using characters such as `š`, `č`, `ž`, `ö` in text (within a `<div>`), they appear and work normally, but do not wrap as they should - instead they break the boundary as if they are actually taking up only about half their width.

This colorful example should demonstrate:
![temp](https://cloud.githubusercontent.com/assets/1211672/13259113/49e93368-da56-11e5-93e2-d97f078f410b.PNG)

Above screenshot was taken with `word-wrap: break-word;`, but without it the situation is the same:
![temp](https://cloud.githubusercontent.com/assets/1211672/13259459/b8bfde4e-da57-11e5-9d14-a30a81e44598.PNG)

`mb_convert_encoding` does not affect the outcome.

I would guess it's an issue with generated font metrics, but I can't really tell.

**Comment by bsweeney** [comment]
Can you supply a sample HTML document? So far I have been unable to reproduce the issue.

**Comment by bsweeney** [comment]
Yikes. Not sure what's changes between 0.6 and 0.7 to cause this. Was about to release the final 0.7 but I think we'll need to fix this first.

**Comment by bsweeney** [comment]
Might be a problem in specific versions of PHP. The issue does not appear to be present when using PHP 5.6.

**Comment by sasosmolej** [comment]
I will try a newer PHP version as well, shouldn't take too long.

**Comment by sasosmolej** [comment]
PHP 5.5.30 throws this warning on `render()`:

`Warning:  DOMNode::cloneNode(): ID header already defined in \dompdf\src\FrameDecorator\AbstractFrameDecorator.php on line 148`

Not 100% sure it's not just my bizzaro code, but hopefully it narrows it down.

**Comment by bsweeney** [comment]
I think the `cloneNode` issue is unrelated, see #1027.

I did run across another error while looking into this. I think it's probably unrelated, but until I do more research I'll leave the info here. In PHP 5.6 if I render the following:

```html
<div style="font-family: dejavu sans;">šššššššššššš ššššššššššššššššššššššššššššššššššššššššššššššš ššššššššššššššššššššššššššššššššš šššššššššššššššššššššššššš šššššššššššššššššššššššššš šššššššššššššššššššššššššš šššššššššššššššššššššššššš šššššššššššššššššššššššššš šššššššššššššššššššššššššš šššššššššššššššššš</div>
```

I see the following error:
> PHP Catchable fatal error:  Argument 1 passed to Dompdf\FrameDecorator\AbstractFrameDecorator::copy() must be an instance of DOMNode, boolean given, called in /dompdf/dompdf_git_develop/src/FrameDecorator/Text.php on line 156 and defined in /dompdf/dompdf_git_develop/src/FrameDecorator/AbstractFrameDecorator.php on line 133

**Comment by bsweeney** [comment]
When we took out the dompdf_config.inc.php file for 0.7.0 we forgot to put the `mb_internal_encoding('UTF-8')` call, which ensures that MBString functions parse text as UTF-8 encoded string, elsewhere in the library. The internal encoding in versions of PHP prior to 5.6 is iso-8859-1 by default (?). Handling strings with the wrong encoding could result in, for example, mis-calculating string length when multi-byte characters are present.

With PHP 5.6 the internal encoding is UTF-8 by default, which is why this isn't a problem in that release.

I added the `mb_internal_encoding()` statement to the dompdf constructor which should address the issue sufficiently for the upcoming release. And as more systems upgrade to PHP 5.6 this will be less of a problem. Still, we should additionally update the mb_* function calls to specify the encoding where accepted. This will prevent issues for some edge cases such as installs configured to use an encoding other than UTF-8 or user scripts that switch the internal encoding.

**Comment by bsweeney** [comment]
FYI, for a temporary fix you can just add `mb_internal_encoding('UTF-8');` somewhere in your PHP code prior to using dompdf.

**Comment by bsweeney** [comment]
The fatal error I [noted earlier](#issuecomment-187880503) also appears to be addressed by setting the internal encoding. However, I think we need to look into this a bit more. Even if the character count is off we shouldn't be passing in a boolean to the copy() method.

**Comment by MarcosBL** [comment]
Can confirm adding ``mb_internal_encoding()`` anywhere before to using dompdf fixes this in PHP 5.5.9-1ubuntu4.14
